crates/ruma-federation-api/src/membership.rs

```diff
@@ -1,7 +1,175 @@
 //! Room membership endpoints.
 
+use ruma_common::serde::{from_raw_json_value, JsonCastable, JsonObject, Raw};
+use ruma_events::{AnyStrippedStateEvent, PossiblyRedactedStateEventContent, StrippedStateEvent};
+use serde::{Deserialize, Serialize};
+use serde_json::value::RawValue as RawJsonValue;
+
 pub mod create_invite;
 pub mod create_join_event;
+pub mod create_knock_event;
 pub mod create_leave_event;
 pub mod prepare_join_event;
+pub mod prepare_knock_event;
 pub mod prepare_leave_event;
+
+/// Possible formats of events that may appear in place of stripped state events.
+#[derive(Clone, Debug, Serialize)]
+#[serde(untagged)]
+#[cfg_attr(not(ruma_unstable_exhaustive_types), non_exhaustive)]
+pub enum RawStrippedState {
+    /// A stripped state event.
+    Stripped(Raw<AnyStrippedStateEvent>),
+
+    /// A full federation PDU.
+    #[cfg(feature = "unstable-msc4311")]
+    Pdu(Box<RawJsonValue>),
+}
+
+impl<'de> Deserialize<'de> for RawStrippedState {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: ::serde::Deserializer<'de>,
+    {
+        let json = Box::<RawJsonValue>::deserialize(deserializer)?;
+
+        #[cfg(feature = "unstable-msc4311")]
+        {
+            use serde::de;
+
+            #[derive(Deserialize)]
+            struct PotentialPduDeHelper {
+                auth_events: Option<de::IgnoredAny>,
+                prev_events: Option<de::IgnoredAny>,
+                signatures: Option<de::IgnoredAny>,
+                hashes: Option<de::IgnoredAny>,
+            }
+
+            let PotentialPduDeHelper { auth_events, prev_events, signatures, hashes } =
+                from_raw_json_value(&json)?;
+
+            if auth_events.is_some()
+                && prev_events.is_some()
+                && signatures.is_some()
+                && hashes.is_some()
+            {
+                return Ok(Self::Pdu(json));
+            }
+        }
+
+        from_raw_json_value(&json).map(Self::Stripped)
+    }
+}
+
+impl From<Raw<AnyStrippedStateEvent>> for RawStrippedState {
+    fn from(value: Raw<AnyStrippedStateEvent>) -> Self {
+        Self::Stripped(value)
+    }
+}
+
+impl JsonCastable<RawStrippedState> for AnyStrippedStateEvent {}
+
+impl<C> JsonCastable<RawStrippedState> for StrippedStateEvent<C> where
+    C: PossiblyRedactedStateEventContent
+{
+}
+
+impl JsonCastable<JsonObject> for RawStrippedState {}
```

Comment:
`RawStrippedState` won't be wrapped into `Raw` (in theory) so there is no need to implement the `JsonCastable` trait.